server: token-based Add/Join functionality

aaron-crl opened the pull request:
Draft for add/join handshake development.

Part of #60632

Release justification: None
Release note: None

Commit:
server: initial connection flow for add/join client server complete

* added new protobuf message for add/join protocol
* added CLI command for testing add/join protocol
* added authentication bypass for add/join GRPC endpoint
* added GRP client and server functions for add/join protocol
* removed config dependcy from cert init code where SSLCertDir was fine
* added add/join protocol specific dial options helper in rpc package

Release justification: low, completing functionality behind separate cli command
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/61265)
<!-- Reviewable:end -->

aaron-crl reviewed (commented):
@itsbilal , I'll rebase this off of #61505 or #61512 once they land.

Otherwise I've squashed things to make things easier on both of us.

I've got stubs which can be replaced with the functions you've already written so feel free to take a look but I'll need to do that rebase before this code settles.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61265#-:-MUzSRcZ-Q0wToKgM_Xe:b-am0b56)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal and @knz)



<!-- Sent from Reviewable.io -->

Commit:
server: updated to connect to rebase from #61505

* removed stubs and dummy functions
* cleaned up TODOs

Release justification: low risk, feature flagged functional enhancement
Release note: None

aaron-crl reviewed (commented):
Rebased on top of the current version of #61512 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61265#-:-MV089YYCRgIcJPxDEdz:b-yorf41)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal and @knz)



<!-- Sent from Reviewable.io -->

Commit:
removed gossip check and re-added stubs for testing

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=61265) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=61265) before we can accept your contribution.<br/><hr/>**Aaron Blum** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=61265) it.</sub>
